dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/RedshiftClusterUsageMetricsTask.java

```diff
@@ -0,0 +1,196 @@
+        sink,
+        clusters.stream()
+            .flatMap(
+                cluster -> {
+                  return zipMetricStreams(
+                          metrics.stream()
+                              .map(
+                                  metricConfig -> {
+                                    return getMetricDataPoints(
+                                        cluster.getClusterIdentifier(), metricConfig);
+                                  })
+                              .collect(toImmutableList()))
+                      .map(
+                          metricData ->
+                              Stream.of(new Object[] {cluster.getClusterIdentifier()}, metricData)
+                                  .flatMap(Stream::of)
+                                  .toArray());
+                }));
+  }
+
+  private List<Cluster> listClusters() {
+    AmazonRedshift client = redshiftApiClient();
+    return client.describeClusters(new DescribeClustersRequest()).getClusters();
+  }
+
+  private Stream<Pair<Date, Double>> getMetricDataPoints(
+      String clusterId, MetricConfig metricConfig) {
+    AmazonCloudWatch client = cloudWatchApiClient();
+    GetMetricStatisticsRequest request =
+        new GetMetricStatisticsRequest()
+            .withMetricName(metricConfig.name().name())
+            .withStatistics(metricConfig.type().name())
+            .withNamespace(REDSHIFT_NAMESPACE)
+            .withDimensions(new Dimension().withName("ClusterIdentifier").withValue(clusterId))
+            .withStartTime(Date.from(interval.getStartUTC().toInstant()))
+            .withEndTime(Date.from(interval.getEndExclusiveUTC().toInstant()))
+            .withPeriod(metricDataPeriod());
+
+    GetMetricStatisticsResult result = client.getMetricStatistics(request);
+    return result.getDatapoints().stream()
+        .map(
+            datapoint ->
+                Pair.of(datapoint.getTimestamp(), getDatapointValue(metricConfig, datapoint)));
+  }
+
+  private Double getDatapointValue(MetricConfig metricConfig, Datapoint datapoint) {
+    switch (metricConfig.type()) {
+      case Maximum:
+        return datapoint.getMaximum();
+      case Average:
+        return datapoint.getAverage();
+      default:
+        return null;
+    }
+  }
+
+  private Stream<Object[]> zipMetricStreams(List<Stream<Pair<Date, Double>>> metricStreams) {
+    Map<String, Double[]> metricsMap = new HashMap<String, Double[]>();
+
+    for (int i = 0; i < metricStreams.size(); i++) {
+      Integer metricIndex = Integer.valueOf(i);
+      metricStreams
+          .get(metricIndex)
+          .forEach(
+              dataPoint -> {
+                String dateString = DATE_FORMAT.format(dataPoint.getKey().toInstant());
+                if (!metricsMap.containsKey(dateString)) {
+                  metricsMap.put(dateString, new Double[metricStreams.size()]);
+                }
+                metricsMap.get(dateString)[metricIndex] = dataPoint.getValue();
```

Comment:
I get rid of `Object/Object[]`, Streams and improved a few things in the code. I think it is much more readable now. Serialization to `Object[]` happens in `doRun()` method now that is writing it to CSV file. `zipMetrics` is still based on the index, but I think this is the easiest solution that we can have. Introducing additional maps would help, but I think it will overcomplicate it.

I thought about exporting each metric as a separate collection of CSV files (to avoid metric aggregation), but I think current solution allows us to easily extend data with additional metrics and keep the same ingestion process.